File: android/app/src/main/java/website/msdnna/tessera/util/Recurrence.kt
package website.msdnna.tessera.util

import java.util.Calendar
import website.msdnna.tessera.data.model.Recurrence

/**
 * Client-side mirror of the backend `internal/recur` advance logic, used only to
 * preview a recurrence on the due-date calendar (highlight upcoming occurrences).
 * The backend remains the source of truth. Works in the device's local zone, on
 * epoch-millis instants (matching the due-date picker).
 */

private fun cal(millis: Long): Calendar = Calendar.getInstance().apply { timeInMillis = millis }

private fun dayKeyOf(c: Calendar): String =
    "%04d-%02d-%02d".format(c.get(Calendar.YEAR), c.get(Calendar.MONTH) + 1, c.get(Calendar.DAY_OF_MONTH))

fun millisDayKey(millis: Long): String = dayKeyOf(cal(millis))

private fun daysInMonth(year: Int, month: Int): Int =
    Calendar.getInstance().apply {
        clear()
        set(year, month, 1)
    }.getActualMaximum(Calendar.DAY_OF_MONTH)

private fun dateOn(base: Calendar, year: Int, month: Int, day: Int): Long {
    val d = minOf(day, daysInMonth(year, month))
    return Calendar.getInstance().apply {
        clear()
        set(year, month, d, base.get(Calendar.HOUR_OF_DAY), base.get(Calendar.MINUTE), 0)
    }.timeInMillis
}

private fun applySkip(millis: Long, skip: Boolean): Long {
    if (!skip) return millis
    val c = cal(millis)
    return when (c.get(Calendar.DAY_OF_WEEK)) {
        Calendar.SATURDAY -> millis + 2 * 86_400_000L
        Calendar.SUNDAY -> millis + 86_400_000L
        else -> millis
    }
}

/** weekday 0=Sun..6=Sat (matches the rule's encoding). */
private fun weekday0(c: Calendar): Int = c.get(Calendar.DAY_OF_WEEK) - Calendar.SUNDAY

private fun nextWeekly(from: Calendar, rule: Recurrence): Long {
    val days = rule.weekdays?.toSortedSet() ?: sortedSetOf()
    val iv = maxOf(1, rule.interval)
    if (days.isEmpty()) {
        return (from.clone() as Calendar).apply { add(Calendar.DAY_OF_MONTH, 7 * iv) }.timeInMillis
    }
    // Later selected weekday this Monday-week?
    val monday = (from.clone() as Calendar).apply {
        add(Calendar.DAY_OF_MONTH, -((weekday0(from) + 6) % 7))
    }
    for (i in 1..6) {
        val d = (from.clone() as Calendar).apply { add(Calendar.DAY_OF_MONTH, i) }
        val weekEnd = (monday.clone() as Calendar).apply { add(Calendar.DAY_OF_MONTH, 7) }
        if (d.before(weekEnd) && days.contains(weekday0(d))) return d.timeInMillis
    }
    val target = (monday.clone() as Calendar).apply { add(Calendar.DAY_OF_MONTH, 7 * iv) }
    for (i in 0..6) {
        val d = (target.clone() as Calendar).apply { add(Calendar.DAY_OF_MONTH, i) }
        if (days.contains(weekday0(d))) return d.timeInMillis
    }
    return (from.clone() as Calendar).apply { add(Calendar.DAY_OF_MONTH, 7 * iv) }.timeInMillis
}

/** Next occurrence after [fromMillis], or null when the recurrence has ended. */
fun nextOccurrence(rule: Recurrence?, fromMillis: Long): Long? {
    if (rule == null || rule.freq.isBlank()) return null
    val from = cal(fromMillis)
    val iv = maxOf(1, rule.interval)
    return when (rule.freq) {
        "daily" -> applySkip((from.clone() as Calendar).apply { add(Calendar.DAY_OF_MONTH, iv) }.timeInMillis, rule.skipWeekends)

        "weekly" -> applySkip(nextWeekly(from, rule), rule.skipWeekends)

        "monthly" -> {
            val total = from.get(Calendar.MONTH) + iv
            val y = from.get(Calendar.YEAR) + Math.floorDiv(total, 12)
            val m = Math.floorMod(total, 12)
            dateOn(from, y, m, from.get(Calendar.DAY_OF_MONTH))
        }

        "yearly" -> dateOn(from, from.get(Calendar.YEAR) + iv, from.get(Calendar.MONTH), from.get(Calendar.DAY_OF_MONTH))

        "custom" -> {
            val key = dayKeyOf(from)
            rule.dates?.sorted()?.firstOrNull { it > key }?.let { s ->
                val (yy, mm, dd) = s.split("-").map { it.toInt() }
                Calendar.getInstance().apply {
                    clear()
                    set(yy, mm - 1, dd, from.get(Calendar.HOUR_OF_DAY), from.get(Calendar.MINUTE), 0)
                }.timeInMillis
            }
        }

        else -> null
    }
}

/**
 * Up to [n] upcoming occurrence day-keys (yyyy-MM-dd) starting at [fromMillis].
 * A one-off rule (`once`) shows just the current due + the single next occurrence.
 */
fun occurrenceKeys(rule: Recurrence?, fromMillis: Long?, n: Int = 24): Set<String> {
    val keys = mutableSetOf<String>()
    if (fromMillis == null || rule == null || rule.freq.isBlank()) return keys
    if (rule.freq == "custom") {
        rule.dates?.forEach { keys.add(it) }
        return keys
    }
    val limit = if (rule.once) 1 else n
    var cur: Long = fromMillis
    keys.add(millisDayKey(cur))
    repeat(limit) {
        val nx = nextOccurrence(rule, cur) ?: return keys
        keys.add(millisDayKey(nx))
        cur = nx
    }
    return keys
}

File: frontend/src/utils/recurrence.js
// Client-side mirror of the backend `internal/recur` advance logic, used to
// preview a recurrence on the calendar (highlight upcoming occurrences) and to
// label rules. The backend remains the source of truth; this only previews.
import { i18n } from '@/i18n'

// The option lists are functions, not arrays: a module-level array would be
// built once at import and keep the language of the first render forever
// (pitfall 1 of #2799). `values` stays a constant — it is the wire format.
export const FREQ_VALUES = ['', 'daily', 'weekly', 'monthly', 'yearly', 'custom']
export const TRIGGER_VALUES = ['complete', 'column', 'schedule']

export function freqOptions() {
  return FREQ_VALUES.map((value) => ({
    label: i18n.global.t(`task.recur.freq.${value || 'none'}`),
    value,
  }))
}

export function triggerOptions() {
  return TRIGGER_VALUES.map((value) => ({
    label: i18n.global.t(`task.recur.trigger.${value}`),
    value,
  }))
}

// "каждые 3 недели" — the unit half, declined for the count. Russian needs three
// forms and the accusative ("каждые 2 неделИ"), which ICU pluralisation in the
// catalogue handles; English gets its own two.
export function unitLabel(freq, n) {
  if (!FREQ_VALUES.includes(freq) || freq === '' || freq === 'custom') return ''
  return i18n.global.t(`task.recur.unit.${freq}`, Math.round(Number(n) || 0))
}

const dayKey = (d) =>
  `${d.getFullYear()}-${String(d.getMonth() + 1).padStart(2, '0')}-${String(d.getDate()).padStart(2, '0')}`

const daysInMonth = (y, m) => new Date(y, m + 1, 0).getDate()

// Build a Date on y/m/d (d clamped to the month length), carrying base's time.
function dateOn(base, y, m, d) {
  const last = daysInMonth(y, m)
  const day = Math.min(d, last)
  return new Date(y, m, day, base.getHours(), base.getMinutes(), base.getSeconds(), 0)
}

function applySkip(d, skipWeekends) {
  if (!skipWeekends) return d
  const wd = d.getDay() // 0=Sun..6=Sat
  if (wd === 6) return new Date(d.getTime() + 2 * 86400000)
  if (wd === 0) return new Date(d.getTime() + 86400000)
  return d
}

function mondayOf(d) {
  const offset = (d.getDay() + 6) % 7 // Mon=0..Sun=6
  const m = new Date(d)
  m.setDate(d.getDate() - offset)
  return m
}

function nextWeekly(from, rule) {
  const days = (rule.weekdays || []).slice().sort((a, b) => a - b)
  if (!days.length) {
    const n = new Date(from)
    n.setDate(from.getDate() + 7 * (rule.interval || 1))
    return n
  }
  const set = new Set(days)
  const monday = mondayOf(from)
  for (let i = 1; i < 7; i++) {
    const c = new Date(from)
    c.setDate(from.getDate() + i)
    if (
      c < new Date(monday.getFullYear(), monday.getMonth(), monday.getDate() + 7) &&
      set.has(c.getDay())
    ) {
      return c
    }
  }
  const target = new Date(monday)
  target.setDate(monday.getDate() + 7 * (rule.interval || 1))
  for (let i = 0; i < 7; i++) {
    const c = new Date(target)
    c.setDate(target.getDate() + i)
    if (set.has(c.getDay())) return c
  }
  const n = new Date(from)
  n.setDate(from.getDate() + 7 * (rule.interval || 1))
  return n
}

// nextOccurrence returns the Date after `from` per the rule, or null when the
// recurrence has ended (custom dates exhausted / no rule).
export function nextOccurrence(rule, from) {
  if (!rule || !rule.freq) return null
  const iv = Math.max(1, rule.interval || 1)
  switch (rule.freq) {
    case 'daily': {
      const n = new Date(from)
      n.setDate(from.getDate() + iv)
      return applySkip(n, rule.skip_weekends)
    }
    case 'weekly':
      return applySkip(nextWeekly(from, { ...rule, interval: iv }), rule.skip_weekends)
    case 'monthly': {
      const total = from.getMonth() + iv
      const y = from.getFullYear() + Math.floor(total / 12)
      const m = ((total % 12) + 12) % 12
      return dateOn(from, y, m, rule.day || from.getDate())
    }
    case 'yearly': {
      const y = from.getFullYear() + iv
      const m = rule.month ? rule.month - 1 : from.getMonth()
      return dateOn(from, y, m, rule.day || from.getDate())
    }
    case 'custom': {
      const key = dayKey(from)
      for (const s of (rule.dates || []).slice().sort()) {
        if (s > key) {
          const [yy, mm, dd] = s.split('-').map(Number)
          return new Date(yy, mm - 1, dd, from.getHours(), from.getMinutes(), from.getSeconds(), 0)
        }
      }
      return null
    }
    default:
      return null
  }
}

// occurrences returns up to `n` date keys (YYYY-MM-DD) of upcoming occurrences,
// starting at `fromTs` itself — for highlighting the calendar. A one-off rule
// (`once`) shows just the current due + the single next occurrence.
export function occurrenceKeys(rule, fromTs, n = 24) {
  const keys = new Set()
  if (!fromTs || !rule || !rule.freq) return keys
  if (rule.freq === 'custom') {
    for (const s of rule.dates || []) keys.add(s)
    return keys
  }
  const limit = rule.once ? 1 : n
  let cur = new Date(fromTs)
  keys.add(dayKey(cur))
  for (let i = 0; i < limit; i++) {
    const nx = nextOccurrence(rule, cur)
    if (!nx) break
    keys.add(dayKey(nx))
    cur = nx
  }
  return keys
}
